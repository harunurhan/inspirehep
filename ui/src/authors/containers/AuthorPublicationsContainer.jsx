import React, { useMemo } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

import LiteratureSearchContainer from '../../literature/containers/LiteratureSearchContainer';
import { AUTHOR_PUBLICATIONS_NS } from '../../reducers/search';

function AuthorPublications({ authorFacetName }) {
  const baseQuery = useMemo(
    () => ({
      author: [authorFacetName],
    }),
    [authorFacetName]
  );
  const baseAggregationsQuery = useMemo(
    () => ({
      author_recid: authorFacetName,
    }),
    [authorFacetName]
  );

  return (
    <LiteratureSearchContainer
      namespace={AUTHOR_PUBLICATIONS_NS}
      baseQuery={baseQuery}
      baseAggregationsQuery={baseAggregationsQuery}
      noResultsTitle="0 Research works"
      embedded
    />
  );
}

AuthorPublications.propTypes = {
  authorFacetName: PropTypes.string.isRequired,
};

const stateToProps = state => ({
  authorFacetName: state.authors.getIn([
    'data',
    'metadata',
    'facet_author_name',
  ]),
});

export default connect(stateToProps)(AuthorPublications);
