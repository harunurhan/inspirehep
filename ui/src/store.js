import { applyMiddleware, createStore } from 'redux';
import thunk from 'redux-thunk';
import createHistory from 'history/createBrowserHistory';
import { routerMiddleware } from 'react-router-redux';
/* eslint-disable import/no-extraneous-dependencies */
import { createLogger } from 'redux-logger';
/* eslint-disable import/no-extraneous-dependencies */
import { composeWithDevTools } from 'redux-devtools-extension';

import reducers from './reducers';
import http from './common/http';
import queryParamsParserMiddleware from './middlewares/queryParamsParser';
import keepPreviousUrlMiddleware from './middlewares/keepPreviousUrl';
import {
  reHydrateRootStateFromStorage,
  createPersistToStorageMiddleware,
} from './middlewares/statePersister';
import searchDispatcherMiddleware from './middlewares/searchDispatcher';
import redirectToErrorPageMiddleware from './middlewares/redirectToErrorPage';
import actionTrackerMiddleware from './middlewares/actionTracker';

export const thunkMiddleware = thunk.withExtraArgument(http);

export const history = createHistory();
const reduxRouterMiddleware = routerMiddleware(history);

const reducersToPersist = ['ui', 'user'];

const PROD_MIDDLEWARES = [
  reduxRouterMiddleware,
  queryParamsParserMiddleware,
  keepPreviousUrlMiddleware,
  createPersistToStorageMiddleware(reducersToPersist),
  searchDispatcherMiddleware,
  redirectToErrorPageMiddleware,
  actionTrackerMiddleware,
  thunkMiddleware,
];

const DEV_MIDDLEWARES = [...PROD_MIDDLEWARES, createLogger()];

const withMiddlewares = () => {
  if (process.env.NODE_ENV === 'production') {
    return applyMiddleware(...PROD_MIDDLEWARES);
  }
  return applyMiddleware(...DEV_MIDDLEWARES);
};

export const store = createStore(
  reducers,
  reHydrateRootStateFromStorage(reducersToPersist),
  composeWithDevTools(withMiddlewares())
);
