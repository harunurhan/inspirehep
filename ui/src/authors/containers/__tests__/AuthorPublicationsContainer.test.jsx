import React from 'react';
import { mount } from 'enzyme';
import { fromJS } from 'immutable';
import { Provider } from 'react-redux';

import { getStoreWithState } from '../../../fixtures/store';
import AuthorPublicationsContainer from '../AuthorPublicationsContainer';
import EmbeddedSearch from '../../../common/components/EmbeddedSearch';
import {
  AUTHOR_PUBLICATIONS_REQUEST,
  AUTHOR_PUBLICATIONS_FACETS_REQUEST,
} from '../../../actions/actionTypes';

describe('AuthorPublicationsContainer', () => {
  it('passes all props to embedded search', () => {
    const store = getStoreWithState({
      authors: fromJS({
        publications: {
          query: { size: 10, page: 2, q: 'dude', sort: 'mostrecent' },
          results: [{ control_number: 1 }, { control_number: 2 }],
          aggregations: { agg1: { foo: 'bar' } },
          sortOptions: [{ value: 'mostrecent', display: 'Most Recent' }],
          loadingResults: true,
          loadingAggregations: true,
          total: 50,
          error: { message: 'Error' },
        },
      }),
    });
    const wrapper = mount(
      <Provider store={store}>
        <AuthorPublicationsContainer renderResultItem={jest.fn()} />
      </Provider>
    );

    expect(wrapper.find(EmbeddedSearch)).toHaveProp({
      query: { size: 10, page: 2, q: 'dude', sort: 'mostrecent' },
      results: fromJS([{ control_number: 1 }, { control_number: 2 }]),
      aggregations: fromJS({ agg1: { foo: 'bar' } }),
      sortOptions: [{ value: 'mostrecent', display: 'Most Recent' }],
      loadingResults: true,
      loadingAggregations: true,
      numberOfResults: 50,
      error: fromJS({ message: 'Error' }),
    });
  });

  // FIXME: avoid also testing that actions merging newQuery and state query
  it('dispatches fetch author publications and facets onQueryChange', () => {
    const existingQuery = {
      page: 3,
      size: 10,
      author: ['Harun'],
      sort: 'mostrecent',
    };
    const queryChange = { sort: 'mostcited' };
    const newQuery = {
      page: 3,
      size: 10,
      sort: 'mostcited',
      author: ['Harun'],
    };
    const store = getStoreWithState({
      authors: fromJS({
        publications: {
          query: existingQuery,
        },
      }),
    });
    const wrapper = mount(
      <Provider store={store}>
        <AuthorPublicationsContainer renderResultItem={jest.fn()} />
      </Provider>
    );
    wrapper.find(EmbeddedSearch).prop('onQueryChange')(queryChange);

    const expectedActions = [
      {
        type: AUTHOR_PUBLICATIONS_REQUEST,
        payload: newQuery,
      },
      {
        type: AUTHOR_PUBLICATIONS_FACETS_REQUEST,
        payload: newQuery,
      },
    ];
    const actions = store.getActions().slice(0, 3); // slice to assert only REQUEST actions
    expect(actions).toEqual(expectedActions);
  });
});
