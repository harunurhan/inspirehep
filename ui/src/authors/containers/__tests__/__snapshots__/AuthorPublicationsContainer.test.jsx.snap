// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthorPublicationsContainer passes all props to embedded search 1`] = `
<EmbeddedSearch
  aggregations={
    Immutable.Map {
      "agg1": Immutable.Map {
        "foo": "bar",
      },
    }
  }
  error={
    Immutable.Map {
      "message": "Error",
    }
  }
  loadingAggregations={true}
  loadingResults={true}
  numberOfResults={0}
  onQueryChange={[Function]}
  query={
    Object {
      "page": 2,
      "q": "dude",
      "size": 10,
      "sort": "mostrecent",
    }
  }
  renderResultItem={[MockFunction]}
  results={
    Immutable.List [
      Immutable.Map {
        "control_number": 1,
      },
      Immutable.Map {
        "control_number": 2,
      },
    ]
  }
  sortOptions={
    Array [
      Object {
        "display": "Most Recent",
        "value": "mostrecent",
      },
    ]
  }
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
  storeSubscription={
    Subscription {
      "listeners": Object {
        "clear": [Function],
        "get": [Function],
        "notify": [Function],
        "subscribe": [Function],
      },
      "onStateChange": [Function],
      "parentSub": undefined,
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "unsubscribe": [Function],
    }
  }
/>
`;
